### PR TITLE
Fixing a bug where an account might think of itself as a peer

### PR DESCRIPTION
**Pull request by andreev-io**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# Xcode
+#
+# gitignore contributors: remember to update Global/Xcode.gitignore, Objective-C.gitignore & Swift.gitignore
+*project.xcworkspace/
+## User settings
+xcuserdata/
+
+## compatibility with Xcode 8 and earlier (ignoring not required starting Xcode 9)
```
Comment (alxdrlitreev) [inline .gitignore]: We don't need it.

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+*.xcscmblueprint
+*.xccheckout
+
+## compatibility with Xcode 3 and earlier (ignoring not required starting Xcode 4)
```
Comment (alxdrlitreev) [inline .gitignore]: We don't need it either.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+*.pbxuser
+!default.pbxuser
+*.mode1v3
+!default.mode1v3
+*.mode2v3
+!default.mode2v3
+*.perspectivev3
+!default.perspectivev3
+
+## Obj-C/Swift specific
```
Comment (alxdrlitreev) [inline .gitignore]: We don't have any Obj-C header mappings to ignore. Unneccessary change, remove it.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+*.hmap
+
+## App packaging
```
Comment (alxdrlitreev) [inline .gitignore]: This is not an iOS app. It can't have neither .ipa nor .dSYM's

File: .gitignore [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+*.ipa
+*.dSYM.zip
+*.dSYM
+
+## Playgrounds
```
Comment (alxdrlitreev) [inline .gitignore]: We're not using Playgrounds, unneccessary change.

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+timeline.xctimeline
+playground.xcworkspace
+
+# Swift Package Manager
```
Comment (alxdrlitreev) [inline .gitignore]: Remove whole section, it's redundant.

File: .gitignore [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+# Package.resolved
+# *.xcodeproj
+#
+# Xcode automatically generates this directory with a .xcworkspacedata file and xcuserdata
+# hence it is not needed unless you have added a package configuration file to your project
+# .swiftpm
+
+.build/
+
+# CocoaPods
```
Comment (alxdrlitreev) [inline .gitignore]: We're not using CocoaPods, remove this unneccessary change.

File: .gitignore [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+# We recommend against adding the Pods directory to your .gitignore. However
+# you should judge for yourself, the pros and cons are mentioned at:
+# https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control
+#
+# Pods/
+#
+# Add this line if you want to avoid checking in source code from the Xcode workspace
+# *.xcworkspace
+
+# Carthage
```
Comment (alxdrlitreev) [inline .gitignore]: We're not using Carthage either. Remove.

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+#
+# Add this line if you want to avoid checking in source code from Carthage dependencies.
+# Carthage/Checkouts
+
+Carthage/Build/
+
+# Accio dependency management
```
Comment (alxdrlitreev) [inline .gitignore]: Accio is not in use in our app. Remove it.

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+Dependencies/
+.accio/
+
+# fastlane
```
Comment (alxdrlitreev) [inline .gitignore]: We have no Fastlane. Remove this section.

File: .gitignore [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+# Instead, use fastlane to re-generate the screenshots whenever they are needed.
+# For more information about the recommended setup visit:
+# https://docs.fastlane.tools/best-practices/source-control/#source-control
+
+fastlane/report.xml
+fastlane/Preview.html
+fastlane/screenshots/**/*.png
+fastlane/test_output
+
+# Code Injection
```
Comment (alxdrlitreev) [inline .gitignore]: It's macOS app, not iOS app. We don't need iOS injections to be in .gitignore.

File: .gitignore [hunk 12]
```diff
@@ -0,0 +1,90 @@
+# Xcode
+#
+# gitignore contributors: remember to update Global/Xcode.gitignore, Objective-C.gitignore & Swift.gitignore
+*project.xcworkspace/
+## User settings
+xcuserdata/
```
Comment (alxdrlitreev) [inline .gitignore]: This is the **only** line that should be in this file.

File: Airtap.xcodeproj/project.pbxproj [hunk 13]
```diff
@@ -540,7 +540,7 @@
 				CODE_SIGN_STYLE = Automatic;
 				COMBINE_HIDPI_IMAGES = YES;
 				DEVELOPMENT_ASSET_PATHS = "\"Airtap/Preview Content\"";
-				DEVELOPMENT_TEAM = A9TLDULDQ2;
+				DEVELOPMENT_TEAM = 6VJL86G6WU;
```
Comment (alxdrlitreev) [inline Airtap.xcodeproj/project.pbxproj]: Unneccessary change, remove it.

File: Airtap/Network/LinkHandler.swift [hunk 14]
```diff
@@ -40,11 +40,13 @@ class LinkHandler: LinkHandling {
             .sink(receiveCompletion: { _ in
                 // no-op
             }, receiveValue: { [weak self] response in
-                self?.persistenceProvider.insertPeer(
-                    id: response.accountId,
-                    firstName: response.firstName,
-                    lastName: response.lastName
-                )
+                if response.accountId != self?.persistenceProvider.selfAccountId {
```
Comment (alxdrlitreev) [inline Airtap/Network/LinkHandler.swift]: Do not use PersistenceProvider to retrieve `selfAccountId`. It's a bad practice. For our auth data we're using `AuthProvider` and data is already presented there. Just use it, inject `authProvider` into `LinkHandler` in Resolver.

File: Airtap/Providers/PersistenceProvider.swift [hunk 15]
```diff
@@ -19,6 +19,7 @@ enum PersistenceProviderEvent: Equatable {
 protocol PersistenceProviding {
     var eventSubject: PassthroughSubject<PersistenceProviderEvent, Never> { get }
     var peers: [Peer] { get }
+    var selfAccountId: Int { get }
```
Comment (alxdrlitreev) [inline Airtap/Providers/PersistenceProvider.swift]: No.

File: Airtap/Providers/PersistenceProvider.swift [hunk 16]
```diff
@@ -32,6 +33,7 @@ class PersistenceProvider: PersistenceProviding {
     private var cancellables = Set<AnyCancellable>()
     
     private(set) var peers: [Peer] = []
+    public var selfAccountId: Int = 0
```
Comment (alxdrlitreev) [inline Airtap/Providers/PersistenceProvider.swift]: wtf it's zero

File: Airtap/Providers/PersistenceProvider.swift [hunk 17]
```diff
@@ -41,14 +43,16 @@ class PersistenceProvider: PersistenceProviding {
...
+                case .signedIn(let id, _):
+                    self?.selfAccountId = id
                     self?.loadPersistence()
                 case .signedOut:
                     self?.destroyPersistence()
                 }
             }
             .store(in: &cancellables)
     }
+
```
Comment (alxdrlitreev) [inline Airtap/Providers/PersistenceProvider.swift:52]: Unneccessary change.

File: Airtap/Providers/PersistenceProvider.swift [hunk 18]
```diff
@@ -41,14 +43,16 @@ class PersistenceProvider: PersistenceProviding {
             .receive(on: DispatchQueue.main)
             .sink { [weak self] event in
                 switch event {
-                case .signedIn:
+                case .signedIn(let id, _):
+                    self?.selfAccountId = id
```
Comment (alxdrlitreev) [inline Airtap/Providers/PersistenceProvider.swift]: Shouldn't be there.

**Review by alxdrlitreev — CHANGES_REQUESTED** [review]
Revert `PersistenceProvider` back and do it with `AuthProvider`.

File: Airtap.xcodeproj/project.pbxproj (hunk 13 above)
Comment (andreev-io) [reply Airtap.xcodeproj/project.pbxproj]: How is this supposed to be done? Xcode changed this for me. I don't want to change it back manually on every push.

**Comment by andreev-io** [comment]
Addressed all comments and pushed changes.

File: Airtap/Providers/PersistenceProvider.swift (hunk 17 above)
Comment (andreev-io) [reply Airtap/Providers/PersistenceProvider.swift:52]: This is actually a good change. The line was unnecessarily tabulated.

File: Airtap.xcodeproj/project.pbxproj (hunk 13 above)
Comment (alxdrlitreev) [reply Airtap.xcodeproj/project.pbxproj]: Commit with Xcode. Just uncheck project file before each commit.

File: Airtap/Providers/AuthProvider.swift [hunk 19]
```diff
@@ -42,6 +44,13 @@ class AuthProvider: AuthProviding {
         return false
     }
     
+    var accountId: Int {
+        if let id = keychain.get("accountId") {
```
Comment (alxdrlitreev) [inline Airtap/Providers/AuthProvider.swift]: Better, but still no. Retrieval from Keychain is quite power-consuming thing. 

Make `accountId` optional and non-computable. Like this:

```
var accountId: Int?
```

Then, populate it with proper value on `signIn()` and set it to `nil` on `signOut()`.

**Review by alxdrlitreev — CHANGES_REQUESTED** [review]
Need bit more improvements, but way better than it was.

File: Airtap/Providers/AuthProvider.swift (hunk 19 above)
Comment (andreev-io) [reply Airtap/Providers/AuthProvider.swift]: Fixed

File: Airtap/Providers/AuthProvider.swift [hunk 20]
```diff
@@ -31,6 +33,8 @@ class AuthProvider: AuthProviding {
     
     private(set) var eventSubject = PassthroughSubject<AuthProviderEvent, Never>()
     
+    var accountId: Int?
```
Comment (alxdrlitreev) [inline Airtap/Providers/AuthProvider.swift]: It shouldn't allow anyone but AuthProvider to write into `accountId`. Let's do last fix. Change it to:
```
private(set) var accountId: Int?
```

**Review by alxdrlitreev — CHANGES_REQUESTED** [review]
Fix `accountId` visibility.

File: Airtap/Providers/AuthProvider.swift (hunk 20 above)
Comment (andreev-io) [reply Airtap/Providers/AuthProvider.swift]: That's neat. Just pushed the changed and squashed the fixup. Feel free to merge.

**Review by alxdrlitreev — APPROVED** [review]
Good job.

**Pull request merged**